evalml/pipelines/components/transformers/preprocessing/polynomial_decomposer.py

```diff
@@ -0,0 +1,410 @@
+        degree: int = 1,
+        seasonal_period: int = -1,
+        random_seed: int = 0,
+        **kwargs,
+    ):
+        def raise_typeerror_if_not_int(var_name, var_value):
+            if not isinstance(var_value, int):
+                if isinstance(var_value, float) and var_value.is_integer():
+                    var_value = int(var_value)
+                else:
+                    raise TypeError(
+                        f"Parameter 'degree' must be an integer!: Received {type(degree).__name__}",
+                    )
+            return var_value
+
+        degree = raise_typeerror_if_not_int("degree", degree)
+        self.seasonal_period = raise_typeerror_if_not_int(
+            "seasonal_period",
+            seasonal_period,
+        )
+
+        params = {"degree": degree, "seasonal_period": self.seasonal_period}
+        params.update(kwargs)
+        error_msg = "sktime is not installed. Please install using 'pip install sktime'"
+
+        trend = import_or_raise("sktime.forecasting.trend", error_msg=error_msg)
+        detrend = import_or_raise(
+            "sktime.transformations.series.detrend",
+            error_msg=error_msg,
+        )
+
+        decomposer = detrend.Detrender(trend.PolynomialTrendForecaster(degree=degree))
+
+        params["time_index"] = time_index
+
+        super().__init__(
+            parameters=params,
+            component_obj=decomposer,
+            random_seed=random_seed,
+        )
+
+    def _set_time_index(self, X, y):
+        """Ensures that target data has a pandas.DatetimeIndex that matches feature data."""
+        dt_df = infer_feature_types(X)
+
+        # Use the feature data's index, preferentially
+        if isinstance(dt_df.index, pd.DatetimeIndex):
+            dt_col = pd.Series(dt_df.index)
+        elif dt_df.ww.select("Datetime").shape[1] == 0:
+            raise ValueError(
+                "There are no Datetime features in the feature data and neither the feature nor the target data have a DateTime index.",
+            )
+        # Use a datetime column of the features if there's only one
+        elif dt_df.ww.select("Datetime").shape[1] == 1:
+            dt_col = dt_df.ww.select("Datetime").squeeze()
+        # With more than one datetime column, use the time_index parameter, if provided.
+        elif dt_df.ww.select("Datetime").shape[1] > 1:
+            if self.parameters.get("time_index", None) is None:
+                raise ValueError(
+                    "Too many Datetime features provided in data but no time_index column specified during __init__.",
+                )
+            elif not self.parameters["time_index"] in X:
+                time_index_col = self.parameters["time_index"]
+                raise ValueError(
+                    f"Too many Datetime features provided in data and provided time_index column {time_index_col} not present in data.",
+                )
+            dt_col = dt_df.ww[self.parameters["time_index"]]
```

Comment:
shouldn't the given `time_index` be the top priority? I suggest refactoring such that we use the given time index first.

```
if self.parameters["time_index"]:
    dt_col = dt_df.ww[self.parameters["time_index"]]=
elif isinstance(dt_df.index, pd.DatetimeIndex):
            dt_col = pd.Series(dt_df.index)
elif dt_df.ww.select("Datetime").shape[1] == 0:
            raise ValueError(
                "There are no Datetime features in the feature data and neither the feature nor the target data have a DateTime index.",
            )
        # Use a datetime column of the features if there's only one
elif dt_df.ww.select("Datetime").shape[1] == 1:
            dt_col = dt_df.ww.select("Datetime").squeeze()
        # With more than one datetime column, use the time_index parameter, if provided.
elif dt_df.ww.select("Datetime").shape[1] > 1:
            if self.parameters.get("time_index", None) is None:
                raise ValueError(
                    "Too many Datetime features provided in data but no time_index column specified during __init__.",
                )
           elif not self.parameters["time_index"] in X:
                time_index_col = self.parameters["time_index"]
                raise ValueError(
                    f"Too many Datetime features provided in data and provided time_index column {time_index_col} not present in data.",
                ) 
```